### PR TITLE
[secure] Introduce vault storage and some underlying fun stuff

### PR DESCRIPTION
**Pull request by davidiw**
The first few PRs set the stage for the real excitement:
1) Fix a bug with LCS deserialization when tags are assigned to enums
2) Add base64 encoding / decoding for storage values
3) Have an availability check and verify it before beginning tests
4) Introduce vault!

This isn't a complete work on Vault, we still need to implement permissions and explicit key management.

### CODE REVIEW

File: secure/storage/src/vault.rs [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+        Ok(())
+    }
+}
+
+impl Storage for VaultStorage {
+    fn available(&self) -> bool {
+        self.client.unsealed().unwrap_or(false)
+    }
+
+    fn create(&mut self, key: &str, value: Value, _permissions: &Permissions) -> Result<(), Error> {
```
Comment (zekun000) [inline secure/storage/src/vault.rs]: the permissions will be used to verify the client.token?

File: docker/vault/run.sh [hunk 2]
```diff
@@ -3,4 +3,10 @@
...
 set -ex
 
-docker run --cap-add=IPC_LOCK --publish 8200:8200 --detach vault
+docker run \
+    --cap-add=IPC_LOCK \
+    -e 'VAULT_DEV_LISTEN_ADDRESS=0.0.0.0:8200' \
+    -e 'VAULT_DEV_ROOT_TOKEN_ID=root_token' \
+    --publish 8200:8200 \
+    --detach \
+    vault
```
[the remark this answers is not in the document]
Comment (davidiw) [reply docker/vault/run.sh]: This is a docker container for vault that @mgorven setup. I was making the parameters explicit and I couldn't find a default root token for dev mode, so I added that.

All this does is wrap around asking docker to install this container and run it.

File: secure/storage/src/vault.rs [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    pub fn reset(&self) -> Result<(), Error> {
+        let secrets = self.client.list_secrets("")?;
+        for secret in secrets {
+            self.client.delete_secret(&secret)?;
+        }
+        Ok(())
+    }
+
+    fn get_secret(&self, key: &str) -> Result<Value, Error> {
+        let value = self.client.read_secret(key, key)?;
```
[the remark this answers is not in the document]
Comment (davidiw) [reply secure/storage/src/vault.rs]: secret is the internal storage framework used by vault. I feel I should add a comment to the top of this file describing vault.

File: secure/storage/src/vault.rs (hunk 1 above)
Comment (davidiw) [reply secure/storage/src/vault.rs]: yes, I have a few more commits not released that actually make sense of this part of the framework :).

**Comment by mgorven** [comment]
What keys get stored here? It may make sense to have a single Vault secret with multiple kv pairs inside.

File: secure/storage/src/vault.rs [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+impl VaultStorage {
+    pub fn new(host: String, token: String) -> Self {
+        Self {
+            client: Client::new(host, token),
+        }
+    }
+
+    /// Erase all secrets and keys from the vault storage. Use with caution.
+    pub fn reset(&self) -> Result<(), Error> {
+        let secrets = self.client.list_secrets("")?;
```
Comment (mgorven) [inline secure/storage/src/vault.rs]: This looks dangerous, I think this should be scoped in someway (e.g. secrets with a configured prefix).

File: secure/storage/vault/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+    fn from(error: serde_json::error::Error) -> Self {
+        Self::SerializationError(format!("{}", error))
+    }
+}
+
+/// Client provides a client around the restful interface to a Vault servce. Learn more
+/// here: https://www.vaultproject.io/api-docs/
+pub struct Client {
+    host: String,
+    token: String,
```
Comment (mgorven) [inline secure/storage/vault/src/lib.rs]: Doesn't need to be done now, but FYI tokens have an expiry and need to be periodically renewed by the client.

File: secure/storage/src/vault.rs [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+    /// Inserts a key, value pair into a secret that shares the name of the key.
+    fn set_secret(&self, key: &str, value: Value) -> Result<(), Error> {
+        self.client.write_secret(key, key, &value.to_base64()?)?;
+        Ok(())
+    }
+}
+
+impl Storage for VaultStorage {
+    fn available(&self) -> bool {
+        self.client.unsealed().unwrap_or(false)
```
Comment (mgorven) [inline secure/storage/src/vault.rs]: I'm not sure if this can typically be queried in prod; the client should probably just use Vault if it is configured.

**Comment by davidiw** [comment]
> What keys get stored here? It may make sense to have a single Vault secret with multiple kv pairs inside.

That's a reasonable consideration. At this point, I didn't want to over pivot on to application / domain specific technologies. The advantage of the current approach is that we define the access policies at the key level granularity. In the other model, we would need to have an interface that allows us to create a backend with permissions.

In the MVP, my intent is to just store all data in here (consensus safety data and private keys). In the future, we'll extend the framework to support generation of keys and signing of data.

**Comment by mgorven** [comment]
Can we add a prefix added to key names then (even if it's hardcoded for now)? While this is fine when running against a dedicated Vault instead in dev mode, it's not gonna fly against a shared prod setup.

File: secure/storage/src/vault.rs (hunk 4 above)
Comment (davidiw) [reply secure/storage/src/vault.rs]: That's an interesting point, my preference at this point would be to make an issue tracking remaining vault issues / deployment constraints. In practice, I believe for V1 most folks will use an application specific vault instance. I can see there being value, however, in namespacing as you suggest and making that accessible via configuration.

File: secure/storage/vault/src/lib.rs (hunk 5 above)
Comment (davidiw) [reply secure/storage/vault/src/lib.rs]: Yes! I realized that too when reading from token API. Considering the complexity here, we should see how hard it is to make an indefinite versus a rotation framework for tokens. Sigh.

File: secure/storage/src/vault.rs (hunk 6 above)
Comment (davidiw) [reply secure/storage/src/vault.rs]: Oh yeah, this is purely for testing purposes to identify why issues might be occurring.

**Comment by davidiw** [comment]
> Can we add a prefix added to key names then (even if it's hardcoded for now)? While this is fine when running against a dedicated Vault in dev mode, it's not gonna fly against a shared prod setup.

Mind if we punt on this, so that I make a proper workstream of this from configuration -> deployment? I'd also like to explore / chat about best practices in these cases.

**Comment by mgorven** [comment]
> Mind if we punt on this, so that I make a proper workstream of this from configuration -> deployment? I'd also like to explore / chat about best practices in these cases.

Okay.

**Pull request closed without merging**